src/data/set/pi.lean

```diff
@@ -0,0 +1,22 @@
+/-
+Copyright (c) 2022 Rishikesh Vaishnav. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rishikesh Vaishnav
+-/
+import data.set
+
+/-!
+# Pi types over sets
+Generalization of `set.restrict` to pi types.
+
+## Main definitions
+* `set.pi_restrict f s` : restrict the domain of `f` to the set `s`;
+-/
+namespace set
+
+/-- Restrict domain of a pi type `f` to a set `s`. Same as `subtype.restrict` but this
+takes the restricting set before the pi type. Generalization of `set.restrict` to pi types. -/
```

Comment:
Can we just generalize the existing `set.restrict` in place? Or does that break everything?

Reply:
Just tried this and there were no issues! However `set.restrict` takes the index set before the pi type, so it doesn't work well for images without a lambda abstraction:

```lean
import data.set.basic

namespace set

def restrict' {α : Type*} {β : α → Type*} (f : Π a : α, β a) (s : set α) : Π a : s, β a := λ x, f x

-- fails
--example {α : Type*} {β : α → Type*} (s : set α) (x : set (Π a : α, β a)) : set (Π a : s, β a) :=
--(s.restrict') '' x

example {α : Type*} {β : α → Type*} (s : set α) (x : set (Π a : α, β a)) : set (Π a : s, β a) :=
(λ x', s.restrict' x') '' x

end set
```

I feel like there's some easy way around this? Or if not should we make something like `set.restrict'` with the order we want?

Reply:
(see #12510)